codes/c/chapter_array_and_linkedlist/array.c

```diff
@@ -16,55 +16,72 @@ int randomAccess(int* nums, int size) {
 }
 
 /* 扩展数组长度 */
-int* extend(int* nums, int size, int enlarge) {
+int* extend(int* nums, int *size, int enlarge) {
+    Assert(enlarge >= 0, "INVALID EXTEND");
     // 初始化一个扩展长度后的数组
-    int* res = (int *)malloc(sizeof(int) * (size + enlarge));
+    // nums is not allocated by malloc()-like function -> don't use realloc() to extend!
+    int* res = (int *)malloc(sizeof(int) * ((*size) + enlarge));
     // 将原数组中的所有元素复制到新数组
-    for (int i = 0; i < size; i++) {
+    for (int i = 0; i < *size; i++) {
         res[i] = nums[i];
     }
     // 初始化扩展后的空间
-    for (int i = size; i < size + enlarge; i++) {
+    for (int i = *size; i < (*size) + enlarge; i++) {
         res[i] = 0;
     }
+    *size += enlarge; 
     // 返回扩展后的新数组
     return res;
 }
 
 /* 在数组的索引 index 处插入元素 num */
-void insert(int* nums, int size, int num, int index) {
+int* insert(int* nums, int *size, int num, int index) {
+    // 0 <= 插入索引 <= size
+    Assert(index >= 0 && index <= *size, "INVALID INDEX");
+    // 插入后需要先加一个位置
+    nums = extend(nums, size , 1);
     // 把索引 index 以及之后的所有元素向后移动一位
-    for (int i = size - 1; i > index; i--) {
+    for (int i = *size - 1; i > index; i--) {
         nums[i] = nums[i - 1];
     }
     // 将 num 赋给 index 处元素
     nums[index] = num;
+
+    return nums;

```

Comment:
Why return `nums`? If it makes sense, why do not return `nums` in `removeItem()`?